File: Submissions/152_Maximum_Product_Subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int result = nums[0];
        int negative = nums[0];
        int positive = nums[0];
        for(int i=1; i<nums.size(); ++i){
            if(nums[i]<0){
                int temp = negative;
                negative = min(nums[i], positive*nums[i]);
                positive = max(nums[i], temp*nums[i]);
            } else {
                negative = min(nums[i], negative*nums[i]);
                positive = max(nums[i], positive*nums[i]);
            }
            result = max(result, positive);
        }
        return result;
    }
};

File: Kotlin/152_Maximum_Product_Subarray.kt
class Solution {
    fun maxProduct(nums: IntArray): Int {
        if(nums.isEmpty()) return 0
        
        var result = nums[0]
        var negative = nums[0]
        var positive = nums[0]
        
        for(i in 1 until nums.size){
            val temp = negative
            
            negative = minOf(positive * nums[i], minOf(negative * nums[i], nums[i]))
            positive = maxOf(positive * nums[i], maxOf(temp * nums[i], nums[i]))

            result = maxOf(result, positive)
        }
        
        return result
    }
}